### PR TITLE
minor ads improvements

### PR DESCRIPTION
**Pull request by ramaraochavali**
Minor ads improvements , refactoring and some unnecessary lock removals (lock only when needed)

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -182,45 +182,45 @@ func (s *DiscoveryServer) receive(con *Connection, reqChannel chan *discovery.Di
...
 func (s *DiscoveryServer) processRequest(req *discovery.DiscoveryRequest, con *Connection) error {
-	if !s.preProcessRequest(con.proxy, req) {
+	if !s.shouldProcessRequest(con.proxy, req) {
 		return nil
 	}
 
-	// For now, don't let xDS piggyback debug requests start watchers
-	if strings.HasPrefix(req.TypeUrl, "istio.io/debug") {
+	// For now, don't let xDS piggyback debug requests start watchers.
+	if req.TypeUrl == v3.DebugType {
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:190]: I think the prefix part is intended

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -298,26 +301,17 @@ func (s *DiscoveryServer) Stream(stream DiscoveryStream) error {
...
-		// on different short-lived go routines started when the push is happening. This would cut in 1/2
-		// the number of long-running go routines, since push is throttled. The main problem is with
-		// closing - the current gRPC library didn't allow closing the stream.
 		select {
-		case req, ok := <-reqChannel:
-			if !ok {
-				// Remote side closed connection or error processing the request.
-				return receiveError
+		case req := <-reqChan:
+			if req != nil {
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:306]: when is req nil?

**Review by howardjohn — COMMENTED** [review]
Looks great

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (adiprerepa) [reply pilot/pkg/xds/ads.go:190]: is there a case where there is stuff after `istio.io/debug`?

Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go:190]: Ah. Ok. Will add it

File: pilot/pkg/xds/ads.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go:314]: it will be nil when channel is closed.

**Comment by esnible** [comment]
@irisdingbj Can you look at this and review it?

**Review by howardjohn — APPROVED** [review]

**Comment by ramaraochavali** [comment]
@irisdingbj PTAL when you get chance

File: pilot/pkg/xds/ads.go [hunk 3]
```diff
@@ -122,55 +129,58 @@ func newConnection(peerAddr string, stream DiscoveryStream) *Connection {
...
+		// Close the initialized channel, if its not already closed, to prevent blocking the stream.
 		select {
 		case <-con.initialized:
 		default:
 			close(con.initialized)
 		}
 	}()
 
-	firstReq := true
+	initialized := false
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:153]: nit: I think firstReq is more clear than initialized

File: pilot/pkg/xds/ads.go [hunk 4]
```diff
@@ -282,13 +292,14 @@ func (s *DiscoveryServer) Stream(stream DiscoveryStream) error {
...
-	var receiveError error
-	reqChannel := make(chan *discovery.DiscoveryRequest, 1)
-	go s.receive(con, reqChannel, &receiveError)
+	// Block until either a request is received or a push is triggered.
+	// We need 2 go routines because 'read' blocks in Recv().
+	//
+	// To avoid 2 routines, we tried to have Recv() in StreamAggregateResource - and the push
+	// on different short-lived go routines started when the push is happening. This would cut in 1/2
+	// the number of long-running go routines, since push is throttled. The main problem is with
+	// closing - the current gRPC library didn't allow closing the stream.
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:301]: Not fully understand the comments

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
I come to realize that when an err happens, we close the reqChannel and errChannel togehther, so in the main stream handling select case, it is no guarantee that the err case can be run into.

**Comment by howardjohn** [comment]
> I come to realize that when an err happens, we close the reqChannel and errChannel together, so in the main stream handling select case, it is no guarantee that the err case can be run into.

I agree,but if we hit the req case it will be nil and we will iterate again and hit the err case(?). Not sure there is a better way?

**Comment by hzxuzhonghu** [comment]
```
If req == nil{
   // recv the err channel.
   ....
}
``` 

and remove the errChannel case may solve this

**Comment by ramaraochavali** [comment]
@hzxuzhonghu changed PTAL

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by ramaraochavali** [comment]
@irisdingbj @istio/wg-user-experience-maintainers can you PTAL?

**Review by shamsher31 — APPROVED** [review]

**Pull request merged**